File: scripts/dev.py
from __future__ import annotations

import os
import signal
import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

try:
    from scripts.kill_ports import format_listener_summary, listener_pids_by_port
except ModuleNotFoundError:
    from kill_ports import format_listener_summary, listener_pids_by_port


@dataclass(frozen=True)
class DevConfig:
    backend_app: str
    backend_host: str
    backend_port: int
    frontend_port: int

    @classmethod
    def from_env(cls) -> "DevConfig":
        return cls(
            backend_app=os.environ.get("BACKEND_APP", "backend.app.main:app"),
            backend_host=os.environ.get("BACKEND_HOST", "127.0.0.1"),
            backend_port=port_from_env("BACKEND_PORT", 8000),
            frontend_port=port_from_env("FRONTEND_PORT", 5173),
        )


def main() -> int:
    config = DevConfig.from_env()
    occupied = listener_pids_by_port([config.backend_port, config.frontend_port])
    if occupied:
        print("Cannot start dev servers because one or more ports are already in use.")
        print(format_listener_summary(occupied))
        print("Run `make kill` to stop existing listeners, or `make restart` to kill and start.")
        print("You can also override BACKEND_PORT or FRONTEND_PORT.")
        return 1

    backend = [
        sys.executable,
        "-m",
        "uvicorn",
        config.backend_app,
        "--reload",
        "--host",
        config.backend_host,
        "--port",
        str(config.backend_port),
    ]
    frontend = [
        sys.executable,
        str(ROOT / "scripts" / "frontend_npm.py"),
        "run",
        "dev",
        "--",
        "--host",
        "127.0.0.1",
        "--port",
        str(config.frontend_port),
    ]

    print(f"Starting backend on http://{config.backend_host}:{config.backend_port}", flush=True)
    backend_process = subprocess.Popen(backend, cwd=ROOT, **process_group_kwargs())
    print(f"Starting frontend on http://127.0.0.1:{config.frontend_port}", flush=True)
    frontend_process = subprocess.Popen(frontend, cwd=ROOT, **process_group_kwargs())
    return wait_for_processes([backend_process, frontend_process])


def port_from_env(name: str, default: int) -> int:
    value = os.environ.get(name)
    if value is None:
        return default
    try:
        port = int(value)
    except ValueError:
        raise SystemExit(f"{name} must be an integer, got {value!r}") from None
    if not 1 <= port <= 65535:
        raise SystemExit(f"{name} must be between 1 and 65535, got {port}")
    return port


def process_group_kwargs() -> dict[str, object]:
    if os.name == "nt":
        return {"creationflags": subprocess.CREATE_NEW_PROCESS_GROUP}
    return {"start_new_session": True}


def wait_for_processes(processes: list[subprocess.Popen[object]]) -> int:
    try:
        while True:
            for process in processes:
                return_code = process.poll()
                if return_code is not None:
                    terminate_all(processes, skip=process)
                    return normalize_return_code(return_code)
            time.sleep(0.25)
    except KeyboardInterrupt:
        terminate_all(processes)
        return 130


def normalize_return_code(return_code: int) -> int:
    if return_code < 0 and -return_code in graceful_exit_signals():
        return 0
    return return_code


def graceful_exit_signals() -> set[int]:
    signals = {signal.SIGTERM, signal.SIGINT}
    if hasattr(signal, "SIGHUP"):
        signals.add(signal.SIGHUP)
    return {int(item) for item in signals}


def terminate_all(
    processes: list[subprocess.Popen[object]],
    *,
    skip: subprocess.Popen[object] | None = None,
) -> None:
    for process in processes:
        if process is skip or process.poll() is not None:
            continue
        terminate_process(process)
    for process in processes:
        if process is skip or process.poll() is not None:
            continue
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()


def terminate_process(process: subprocess.Popen[object]) -> None:
    if os.name == "nt":
        subprocess.run(
            ["taskkill", "/PID", str(process.pid), "/T", "/F"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
        )
        return
    try:
        os.killpg(process.pid, signal.SIGTERM)
    except ProcessLookupError:
        try:
            process.send_signal(signal.SIGTERM)
        except ProcessLookupError:
            return


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/dev.mjs
#!/usr/bin/env node
import { spawn, spawnSync } from "node:child_process";
import { dirname, isAbsolute, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { isPortAvailable } from "./ports.mjs";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const dryRun = process.argv.includes("--dry-run");
const config = {
  npm: process.env.NPM ?? "npm",
  backendDir: resolveFromRoot(process.env.BACKEND_DIR ?? "backend"),
  frontendDir: resolveFromRoot(process.env.FRONTEND_DIR ?? "frontend"),
  backendHost: process.env.BACKEND_HOST ?? "127.0.0.1",
  backendPort: portFromEnv("BACKEND_PORT", 8000),
  frontendPort: portFromEnv("FRONTEND_PORT", 5173),
  backendStaticDir: resolveFromRoot(process.env.BACKEND_STATIC_DIR ?? "backend/static")
};

const children = [];

await main();

async function main() {
  const backend = {
    command: config.npm,
    args: [
      "--prefix",
      config.backendDir,
      "run",
      "dev",
      "--",
      "serve",
      "--host",
      config.backendHost,
      "--port",
      String(config.backendPort),
      "--static-dir",
      config.backendStaticDir
    ]
  };
  const frontend = {
    command: config.npm,
    args: [
      "--prefix",
      config.frontendDir,
      "run",
      "dev",
      "--",
      "--host",
      "127.0.0.1",
      "--port",
      String(config.frontendPort)
    ]
  };

  if (dryRun) {
    console.log(commandLine(backend));
    console.log(commandLine(frontend));
    return;
  }

  const occupied = [];
  if (!(await isPortAvailable(config.backendHost, config.backendPort))) {
    occupied.push(`${config.backendHost}:${config.backendPort}`);
  }
  if (!(await isPortAvailable("127.0.0.1", config.frontendPort))) {
    occupied.push(`127.0.0.1:${config.frontendPort}`);
  }
  if (occupied.length) {
    console.error(`Cannot start dev servers because these address(es) are already in use: ${occupied.join(", ")}`);
    console.error("Run `make kill` to stop existing listeners, or override BACKEND_PORT or FRONTEND_PORT.");
    process.exitCode = 1;
    return;
  }

  console.log(`Starting backend on http://${config.backendHost}:${config.backendPort}`);
  children.push(start(backend));
  console.log(`Starting frontend on http://127.0.0.1:${config.frontendPort}`);
  children.push(start(frontend));
  process.exitCode = await waitForExit(children);
}

function start({ command, args }) {
  return spawn(command, args, {
    cwd: ROOT,
    detached: process.platform !== "win32",
    env: process.env,
    shell: process.platform === "win32",
    stdio: "inherit"
  });
}

async function waitForExit(processes) {
  let shuttingDown = false;
  const shutdown = (signal = "SIGTERM") => {
    if (shuttingDown) {
      return;
    }
    shuttingDown = true;
    for (const child of processes) {
      terminate(child, signal);
    }
  };
  process.once("SIGINT", () => shutdown("SIGINT"));
  process.once("SIGTERM", () => shutdown("SIGTERM"));

  const winner = await Promise.race(
    processes.map(
      (child) =>
        new Promise((resolveExit) => {
          child.once("exit", (code, signal) => resolveExit({ child, code, signal }));
        })
    )
  );
  shutdown();
  await Promise.all(
    processes
      .filter((child) => child !== winner.child && child.exitCode === null)
      .map(
        (child) =>
          new Promise((resolveExit) => {
            child.once("exit", resolveExit);
          })
      )
  );
  return normalizeExitCode(winner.code, winner.signal);
}

function terminate(child, signal) {
  if (child.exitCode !== null || child.pid === undefined) {
    return;
  }
  if (process.platform === "win32") {
    spawnSync("taskkill", ["/PID", String(child.pid), "/T", "/F"], { stdio: "ignore" });
    return;
  }
  try {
    process.kill(-child.pid, signal);
  } catch {
    try {
      process.kill(child.pid, signal);
    } catch {
      // The child may have exited between the status check and the signal.
    }
  }
}

function normalizeExitCode(code, signal) {
  if (typeof code === "number") {
    return code;
  }
  return signal === "SIGINT" || signal === "SIGTERM" ? 0 : 1;
}

function resolveFromRoot(value) {
  return isAbsolute(value) ? value : resolve(ROOT, value);
}

function portFromEnv(name, fallback) {
  const raw = process.env[name];
  if (!raw) {
    return fallback;
  }
  const port = Number.parseInt(raw, 10);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new Error(`${name} must be an integer between 1 and 65535, got ${raw}`);
  }
  return port;
}

function commandLine({ command, args }) {
  return [command, ...args].map(shellQuote).join(" ");
}

function shellQuote(value) {
  return /^[A-Za-z0-9_./:=@-]+$/.test(value) ? value : JSON.stringify(value);
}
